### PR TITLE
CCC page - test cases.

### PR DESCRIPTION
**Pull request by Krishnakumarskr**
The following test cases are covered.

- Hero video should play on click.

- Thumbnail images should be loaded.

- Customer logos should be displayed as per IP.

- Badges should be displayed properly.

- Sneak peek section should display the images properly.

**New functions**

Added two new functions in `simplebrowsers.py`

- `get_from_storage` - Used to get values stored in the local storage by passing key as param.

- `hover` - To perform hover action on web elements.

### CODE REVIEW

File: ccc/test_credit_card.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+import logging
+import pytest
+from common.utils import resize_browser
+
+logger = logging.getLogger(__name__)
+
+@pytest.fixture(scope='function')
+def browser(module_browser, base_url, request):
+    module_browser.get(base_url + '/corporate-credit-cards')
+    time.sleep(3)
```
Comment (aikanshgarg) [inline ccc/test_credit_card.py:13]: I guess we don't require sleeps inside our code now
as siva has added them by default in parent methods

File: ccc/test_credit_card.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    return module_browser
+
+#common utils
+
+def check_spacing(element_1, element_2, space):
+    element_1_Y = element_1.location['y']
```
Comment (aikanshgarg) [inline ccc/test_credit_card.py:19]: can we not use camelCasing?

File: simplebrowser.py [hunk 3]
```diff
@@ -176,4 +178,22 @@ def get_current_url(self):
...
-        self.driver.set_window_size(width, height)
\ No newline at end of file
+        self.driver.set_window_size(width, height)
+
+    def get_from_storage(self, key):
+        return json.loads(self.driver.execute_script("return window.localStorage.getItem(arguments[0]);", key))
+
+    def hover(self, elem, scroll=False):
+        ltag = elem.tag_name.lower() if elem.tag_name else None
+        assert ltag in ['input', 'li', 'button', 'span',
```
Comment (aikanshgarg) [inline simplebrowser.py:188]: why are we allowed to hover on an input element?

File: ccc/test_credit_card.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+@pytest.fixture(scope='function')
+def browser(module_browser, base_url, request):
+    module_browser.get(base_url + '/corporate-credit-cards')
+    time.sleep(3)
+    return module_browser
+
+#common utils
+
+def check_spacing(element_1, element_2, space):
```
Comment (ShihaMohan235) [inline ccc/test_credit_card.py:18]: can we use different casing for variable and functions

File: ccc/test_credit_card.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+def test_ccc_video(browser):
+    play_ccc_hero_video(browser)
+    e = browser.find(xpath="//div[contains(@class, 'feature-hero-video')]//iframe")
+    assert e and e.is_displayed(), 'Video not played'
+
+
+def test_video_thumbnail(browser):
+    browser.find("//div[contains(@class, 'feature-hero-video')]//img", scroll=True)
+    img = browser.driver.find_elements_by_xpath("//div[contains(@class, 'feature-hero-video')]//div[contains(@class, 'youtube')]//img")
```
Comment (ShihaMohan235) [inline ccc/test_credit_card.py:59]: why using find_elements_by_xpath here ?

File: ccc/test_credit_card.py [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+    script = "return(typeof arguments[0].naturalWidth!=\"undefined\" && arguments[0].naturalWidth>0)"
```
Comment (ShihaMohan235) [inline ccc/test_credit_card.py:60]: naturalWidth or width ?

File: ccc/test_credit_card.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+    e = browser.driver.execute_script(script, img[0])
+    assert e, 'Image not loaded'
+
+
+def test_customer_logo(browser):
+    ip_info = browser.get_from_storage('ipInfo')
+    country = ip_info['country']
+
+    if country == 'India':
+        logo_div = browser.find("//div[contains(@class, 'customer-logo-non-india') and contains(@class, 'd-none')]")
```
Comment (ShihaMohan235) [inline ccc/test_credit_card.py:71]: why checking for class-d-none ?

File: ccc/test_credit_card.py [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+        logger.info(logo_div)
+        assert logo_div, 'Found an US image in Indian IP'
+    else:
+        logo_div = browser.find("//div[contains(@class, 'customer-logo-india') and contains(@class, 'd-none')]")
+        assert logo_div, 'Found an Indian image in other IP'
+
+
+def test_badges(browser):
+    total_badges = browser.find_many("//div[contains(@class, 'fyle-badge')]")
+    hidden_badges = browser.find_many("//div[contains(@class, 'fyle-badge') and contains(@style, 'display: none;')]")
```
Comment (ShihaMohan235) [inline ccc/test_credit_card.py:81]: why checking for style ?

File: ccc/test_credit_card.py [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+    assert e, 'Image not loaded'
+
+
+def test_customer_logo(browser):
+    ip_info = browser.get_from_storage('ipInfo')
+    country = ip_info['country']
+
+    if country == 'India':
+        logo_div = browser.find("//div[contains(@class, 'customer-logo-non-india') and contains(@class, 'd-none')]")
+        logger.info(logo_div)
```
Comment (Krishnakumarskr) [inline ccc/test_credit_card.py:72]: Remove logger.

File: ccc/mobile/test_credit_card.py [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+    module_browser.get(base_url + '/corporate-credit-cards')
+    return module_browser
+
+
+def check_spacing(element_1, element_2, space):
+    element_1_Y = element_1.location['y']
+    element_2_Y = element_2.location['y']
+    element_1_height = element_1.size['height']
+    element_2_height = element_2.size['height']
+    space_difference = abs(element_1_Y - element_2_Y) - element_1_height
```
Comment (snarayanank2) [inline ccc/mobile/test_credit_card.py:22]: I don't understand the computation here - what exactly are we trying to do

File: ccc/mobile/test_credit_card.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+	return element.value_of_css_property('margin') == value
+
+
+@pytest.mark.parametrize('browser', [('mobile_1')], indirect=True)
+def test_space_h3_subtext(browser):
+    h3 = browser.find_many("//section[contains(@class, 'long-background')]//div[contains(@class, 'feature-pages-solution')]//h3")
+    subtext = browser.find_many("//section[contains(@class, 'long-background')]//div[contains(@class, 'feature-pages-solution')]//p")
+
+    for i in range(len(h3)):
+        assert check_spacing(h3[i], subtext[i], 10), 'Spacing between h3 and subtext is wrong'
```
Comment (snarayanank2) [inline ccc/mobile/test_credit_card.py:56]: assert message should include the h3 text so it is easy to find out what exactly is broken

File: ccc/mobile/test_credit_card.py [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+
+
+@pytest.mark.parametrize('browser', [('mobile_1')], indirect=True)
+def test_space_travel_benefits(browser):
+    travel_benefits_row = browser.find("//div[contains(@class, 'travel-request-benefits')]")
+    row_padding_top = travel_benefits_row.value_of_css_property('padding-top')
```
Comment (snarayanank2) [inline ccc/mobile/test_credit_card.py:62]: use the padding-top-of function above

File: ccc/mobile/test_credit_card.py [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+def test_space_benefit_text(browser):
+	benefit_texts = browser.find_many("//p[contains(@class, 'benefit-text')]")
+	for text in benefit_texts:
+		assert margin_top_of(text, '5px') and margin_bottom_of(text, '20px'), 'Marign values of benefit text is wrong'
+
+
+@pytest.mark.parametrize('browser', [('mobile_1')], indirect=True)
+def test_space_feature_solution(browser):
+    feature_solution_row = browser.find("//div[contains(@class, 'feature-pages-solution')]")
+    assert padding_top_of(feature_solution_row, '30px'), f'Padding top is incorrect, the correct value is 40px, but {row_padding_top} found'
```
Comment (snarayanank2) [inline ccc/mobile/test_credit_card.py:77]: assert message says 40px but check is for 30px

File: ccc/test_credit_card.py (hunk 2 above)
Comment (snarayanank2) [reply ccc/test_credit_card.py:19]: no please.. let's use `snake_case` - I will put a pylint to enforce casing guidelines.

File: ccc/test_credit_card.py (hunk 4 above)
Comment (snarayanank2) [reply ccc/test_credit_card.py:18]: nope that is not python standard let's stick to snake casing

File: ccc/test_credit_card.py [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+
+def margin_bottom_of(element, value):
+    return element.value_of_css_property('margin-bottom') == value
+
+
+def padding_of(element, value):
+	return element.value_of_css_property('padding') == value
+
+
+def margin_of(element, value):
```
Comment (snarayanank2) [inline ccc/test_credit_card.py:47]: do you think it is valuable to have this (and similar functions be):

assert_margin(element, value) this performs the assert with appropriate message

File: ccc/test_credit_card.py [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+#end of common utils
+
+def play_ccc_hero_video(browser):
+    time.sleep(1)
+    browser.click(xpath="//div[contains(@class, 'play-button')]")
+    time.sleep(5)
+
+
+def test_ccc_video(browser):
+    play_ccc_hero_video(browser)
```
Comment (snarayanank2) [inline ccc/test_credit_card.py:60]: this function only does a click.. inline the code here

File: ccc/test_credit_card.py [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+    e = browser.find(xpath="//div[contains(@class, 'feature-hero-video')]//iframe")
+    assert e and e.is_displayed(), 'Video not played'
+
+
+def test_video_thumbnail(browser):
+    browser.find("//div[contains(@class, 'feature-hero-video')]//img", scroll=True)
```
Comment (snarayanank2) [inline ccc/test_credit_card.py:66]: this is a no-op.. you're not doing anything with the result of the find

File: ccc/test_credit_card.py (hunk 5 above)
Comment (snarayanank2) [reply ccc/test_credit_card.py:67]: just use find instead of find_elements_by_xpath

File: ccc/test_credit_card.py [hunk 17]
```diff
@@ -0,0 +1,255 @@
...
+def test_space_h2_logo(browser):
+    h2 = browser.find("//section[contains(@class, 'customer-logos-v2')]//div[contains(@class, 'col')]//h2")
+    logo = browser.find("//section[contains(@class, 'customer-logos-v2')]//div[contains(@class, 'col')]//div[not(contains(@class, 'd-none'))]")
+    assert check_spacing(h2, logo, 40), 'Spacing between h2 and logo is wrong'
+
+
+def test_cards_url(browser):
+    links = []
+    urls = [
+        'https://ww2.fylehq.com/travel-expense-management',
```
Comment (snarayanank2) [inline ccc/test_credit_card.py:138]: this is not a good pattern.. you should only be testing one page in this file

Comment (snarayanank2) [reply ccc/test_credit_card.py:138]: also base_url should not be hardcoded - we might want to run tests against localhost, staging or production

File: simplebrowser.py [hunk 18]
```diff
@@ -176,4 +178,22 @@ def get_current_url(self):
...
 
     def set_window_size(self, width, height):
-        self.driver.set_window_size(width, height)
\ No newline at end of file
+        self.driver.set_window_size(width, height)
+
+    def get_from_storage(self, key):
+        return json.loads(self.driver.execute_script("return window.localStorage.getItem(arguments[0]);", key))
+
+    def hover(self, elem, scroll=False):
```
Comment (snarayanank2) [inline simplebrowser.py:186]: scroll is not being used

File: simplebrowser.py [hunk 19]
```diff
@@ -176,4 +178,22 @@ def get_current_url(self):
...
+        ltag = elem.tag_name.lower() if elem.tag_name else None
+        assert ltag in ['input', 'li', 'button', 'span',
+                        'a', 'div', 'textarea'], 'xpath did not return proper element'
+        actions = ActionChains(self.driver)
```
Comment (snarayanank2) [inline simplebrowser.py:190]: this is nicely done

File: simplebrowser.py [hunk 20]
```diff
@@ -176,4 +178,22 @@ def get_current_url(self):
...
+        actions.move_to_element(elem)
+        actions.perform()
+        return elem
+
+    def back(self):
+        return self.driver.back()
+
+    def find_by_css(self, css):
```
Comment (snarayanank2) [inline simplebrowser.py:198]: let's use only xpath to find elements so that it is standardized

File: ccc/test_credit_card.py [hunk 21]
```diff
@@ -0,0 +1,240 @@
...
+        '/expense-management'
+    ]
+    cards = browser.find_many("//section[contains(@class, 'explore-fyle-beyond')]//div[contains(@class, 'feature-blocks')]//a")
+
+    for i, card in enumerate(cards):
+        links.append(card.get_attribute('href'))
+
+    for i, link in enumerate(links):
+        browser.get(link)
+        assert (base_url + urls[i]) == browser.get_current_url()
```
Comment (Krishnakumarskr) [inline ccc/test_credit_card.py:142]: add a message

**Pull request closed without merging**